### DISCUSSION TITLE
CWEB parser API

### DISCUSSION THREAD

**Discussion by zmajeed** [gogotrango/cwebdraftapi#16] [General]
A first draft of the CWEB parser API is in https://github.com/gogotrango/cwebdraftapi/tree/main/cwebparser

The C++ API header file is https://github.com/gogotrango/cwebdraftapi/blob/main/cwebparser/src/parserapi/v4/cwebparser.v4.h

There's a `CWebParser` class in the `cweb` namespace - actually in `cweb::v4` namespace while I iterate through draft versions

The API comprises callbacks corresponding to CWEB control codes and document structure

The callbacks are listed in https://github.com/gogotrango/cwebdraftapi/blob/main/cwebparser/src/parserapi/v4/cwebparser_callbacks.v4.h - it has a vector for each type of callback - so multiple callbacks for each type may be registered

There is an example application in https://github.com/gogotrango/cwebdraftapi/tree/main/cwebparser/src/examples/echo - it simply echoes a cweb file to stdout by printing the original text from each call of the callbacks

The echo callbacks are in https://github.com/gogotrango/cwebdraftapi/blob/main/cwebparser/src/examples/echo/echo_cweb_callbacks.h - not all of them are filled in yet because I have only tried it on a few files - mainly `miles_span.w` from Stanford GraphBase - https://github.com/gogotrango/cwebdraftapi/blob/main/testfiles/miles_span.w

The API is pretty simple as is typical for parsing APIs - you write your callbacks - put them in a struct - any struct - and pass the object to `CWebParser::parse` along with the input stream - you don't need to write a callback for parts of the document that don't interest you - currently only one callback of each type can be added but as mentioned above `CWebParser` has an array for each callback type so I do plan to support multiple callbacks for each type of event

Each callback has some parameters specific to the event, e.g. the `sectionName` callback gets the name of the section and parameters for whether it's a prefix and whether it's a continuation. Each callback also gets a `Context` parameter that has the original source text with line and column info.

Here's a snippet from `echo_cweb`

```c++
struct EchoCallbacks {

  function<bool(const string&, const Context&)> text = [](const string& text, const Context&) {
    cout << text;
    return true;
  };

  function<bool(const string&, const Context&)> starredTexSectionBegin = [](const string&, const Context& context) {
    cout << context.text;
    return true;
  };

  function<bool(const CommentType, const Context&)> commentBegin = [](const CommentType, const Context& context) {
    cout << context.text;
    return true;
  };

  function<bool(const CommentType, const Context&)> commentEnd = [](const CommentType, const Context& context) {
    cout << context.text;
    return true;
  };

  function<bool(const string&, const bool, const bool, const Context&)> namedSectionBegin = [](const string&, const bool, const bool, const Context& context) {
    cout << context.text;
    return true;
  };

  function<bool(const string&, const bool, const bool, const Context&)> sectionName = [](const string&, const bool, const bool, const Context& context) {
    cout << context.text;
    return true;
  };

  function<bool(const StringType, const Context&)> stringBegin = [](const StringType, const Context& context) {
    cout << context.text;
    return true;
  };

  function<bool(const StringType, const Context&)> stringEnd = [](const StringType, const Context& context) {
    cout << context.text;
    return true;
  };

};
```

And here's the call to parse

```c++
// set config property to not escape macro newlines
CWebParser cweb({false});
EchoCallbacks echoCallbacks;
cweb.parse(file, echoCallbacks);
```

The API wraps a bison LALR(1) grammar and parser and a flex lexer

The grammar with actions is in https://github.com/gogotrango/cwebdraftapi/blob/main/cwebparser/src/grammar/cwebbisonparser.v4.bison.y

The code for the actions in the grammar is somewhat separated out but it's much nicer to read just the grammar in https://github.com/gogotrango/cwebdraftapi/blob/main/grammar/cweb.grammar.bison.y unless you're really interested in the nitty gritty - and the point of the API is to allow applications to be written without touching the grammar or the low-level bison parser

The lexer rules file is https://github.com/gogotrango/cwebdraftapi/blob/main/cwebparser/src/grammar/cwebbisonparser.v4.flex.l

There are multiple layers of testing - all unit tests use googletest - there are unit tests for the lexer, for the bison parser, and for the parser API - and finally there are example applications that exercise the API

The project is built with cmake - you say something like

```bash
$ cmake -B build -S . -D BISON=/usr/bin/bison
$ make -C build
```
to build everything - then you can run the tests and try `echo_cweb` on a CWEB file

```bash
$ cwebparser/build/src/examples/echo/echo_cweb -h
Usage: echo_cweb [file]
Echo CWEB file to stdout

file: CWEB file, default is stdin

Examples
echo_cweb < miles_span.w
echo_cweb miles_span.w
```

The path to bison is needed because I use bison 3.7 which is pretty recent and may not be the default version on your system

The repo has submodule dependencies on googletest and fmt - the new C++ format library for printing

The code is C++20 - I've been using gcc 11.1 in cygwin on Windows but any C++20 compiler like clang and vc++ should work as long as bison and flex are available - the test drivers and applications however do use gnu `getopt_long` for commandline options processing for now so that too is needed - none of this is a big issue on Linux or Windows cygwin

Although the API is simple - there's a lot going on backstage as anyone who's familiar with CWEB can imagine - also the bison and flex layers are good use cases of how to get the most out of the C++ support offered by these tools -  I plan to document the design and internals as well - and expand the example applications

**Comment by ascherer** [comment]
Just for the record: **Kubuntu 20.04 LTS** provides

* **cmake 3.16.3** < 3.20
* **bison 3.5.1** < 3.7
* **flex 2.6.4**
* **g++ 9.3.0** (with option `-std=c++2a` at best)

and **macOS 10.15.7 Catalina** provides

* _no_ **cmake** (but version **3.20.5** through `brew install cmake`)
* **bison 2.3** (and version **3.7.6** through `brew install bison`)
* **flex 2.5.35** (and version **2.6.4** through `brew install flex`)
* **g++ clang 12.0.0** (also with option `-std=c++2a` at best).

Good luck!

**Reply by zmajeed** [reply]
Thanks Andreas - I should be able to relax the cmake version but not bison 3.7 and C++20 - it's good packages for newer versions of the tooling are available and hopefully more widespread by the time the API stabilizes - another option for a release is to provide the source files generated by flex and bison to remove the need for the tools

**Comment by zmajeed** [comment]
Added a new `trace_cweb` example that demos location info available in each callback - it prints the line and column for the begin and end locations of the event

The source is at https://github.com/gogotrango/cwebdraftapi/tree/main/cwebparser/src/examples/trace - the callbacks look like

```c++
struct TraceCallbacks {

// utility function
  bool printLineInfo(const string& event, const Context& context) const {
    const auto& [bfilename, bline, bcol] = context.loc.begin;
    const auto& [efilename, eline, ecol] = context.loc.end;
    print("{:<6} {:<5} {:<6} {:<5} {}\n", bline, bcol, eline, ecol, event);
    return true;
  }

  function<bool(const string&, const Context&)> text = [this](const string&, const Context& context) {
    return printLineInfo("text", context);
  };

  function<bool(const Context&)> documentBegin = [this](const Context& context) {
    return printLineInfo("documentBegin", context);
  };

  function<bool(const Context&)> documentEnd = [this](const Context& context) {
    return printLineInfo("documentEnd", context);
  };

};
```
and the driver
```c++
CWebParser cweb;
TraceCallbacks traceCallbacks;

ifstream file(argv[optind]);
cweb.parse(file, traceCallbacks);
```

the full sample output is at https://github.com/gogotrango/cwebdraftapi/blob/main/doc/cwebapi_design/trace_miles_span_w_20210706.txt

```bash
$ trace_cweb miles_span.w
...
1579   1     1579   34    namedSectionBegin
1579   34    1580   17    text
1580   17    1580   19    commentBegin
1580   19    1580   40    text
1580   40    1580   41    innerCBegin
1580   41    1580   43    text
1580   43    1580   44    innerCEnd
1580   44    1580   75    text
1580   75    1580   77    commentEnd
1580   77    1581   1     text
1581   1     1581   44    sectionName
1581   44    1582   1     text
1582   1     1582   69    sectionName
1582   69    1583   1     text
1583   1     1583   51    sectionName
1583   51    1585   1     text
1585   1     1585   1     namedSectionEnd
1585   1     1585   3     unstarredTexSectionBegin
1585   3     1585   64    text
1585   64    1585   65    innerCBegin
1585   65    1585   69    text
1585   69    1585   70    innerCEnd
1585   70    1586   45    text
1586   45    1586   46    innerCBegin
1586   46    1586   50    text
1586   50    1586   51    innerCEnd
1586   51    1587   23    text
1587   23    1587   24    innerCBegin
1587   24    1587   30    text
1587   30    1587   31    innerCEnd
1587   31    1589   1     text
1589   1     1589   1     unstarredTexSectionEnd
1589   1     1589   16    namedSectionBegin
1589   16    1603   1     text
1603   1     1603   1     namedSectionEnd
1603   1     1603   3     unstarredTexSectionBegin
1603   3     1603   3     unstarredTexSectionEnd
1603   3     1603   72    namedSectionBegin
1603   72    1607   1     text
1607   1     1607   1     namedSectionEnd
1607   1     1607   3     unstarredTexSectionBegin
...
```
and the source text
```
$ sed -n '1579,1607p' miles_span.w | nl -ba -v1579
1579  @<Do stage 2 of |cher_tar_kar|@>=
1580  gv=g->vertices; /* the global variable |gv| helps access auxiliary memory */
1581  @<Map all vertices to their index numbers@>;
1582  @<Create the reduced matrix by running through all remaining edges@>;
1583  @<Execute Prim's algorithm on the reduced matrix@>;
1584
1585  @ The vertex-mapping algorithm is $O(n)$ because each non-null |comp| link
1586  is examined at most three times. We set the |comp| field to null
1587  as an indication that |findex| has been set.
1588
1589  @<Map all...@>=
1590  if (s==NULL) s=large_list;
1591  else o,t->rsib=large_list;
1592  for (k=0,v=s;v;o,v=v->rsib,k++) o,v->findex=k;
1593  for (v=g->vertices;v<g->vertices+g->n;v++)
1594    if (o,v->comp) {
1595      for (t=v->comp;o,t->comp;t=t->comp) ;
1596      o,k=t->findex;
1597      for (t=v;o,u=t->comp;t=u) {
1598        o,t->comp=NULL;
1599        o,t->findex=k;
1600      }
1601    }
1602
1603  @ @<Create the reduced matrix by running through all remaining edges@>=
1604  for (j=0;j<lo_sqrt;j++) for (k=0;k<lo_sqrt;k++) o,matx(j,k)=INF;
1605  for (kk=0;s;o,s=s->rsib,kk++) qtraverse(s->pq,note_edge);
1606
1607  @ The |note_edge| procedure ``visits'' every edge in the
```
